### DISCUSSION TITLE
I can't install TIC80 version 1.1 on my Android device and the keyboard arrow keys don't respond.

### DISCUSSION THREAD

**Discussion by kuromame96** [nesbox/TIC-80#2304] [Q&A] [answered]
Hello.

First of all, I would like to state that I am not good at English. It may be difficult to understand the meaning.

I recently decided to install the Pro version of Tic80 version 1.1 from itch.io on my FireHD 8 2018. However, I was not able to install it.
I read on an internet information site that it can be installed using Split APKs Installer.
So, I also tried to see if I could install it using this app.
However, an error message appeared and I was unable to install it.
I got an error message like this.

INSTALL_FAILED_NO_MATCHING_ABIS: Failed to extract native libraries, res=-113

Can you solve this problem?

It was also possible to install it on another Android device.
However, I cannot operate it using the arrow keys on my keyboard.
Arrow keys respond properly in TIC80 version 1.0.
The arrow keys respond normally if I use a different keyboard, but I like the keyboard I'm currently using, so I'd like to see this issue fixed as well.
The model number is ROYAL KLUDGE Model: RK61.
Thank you for your help in fixing the problem.

**Comment by sogaiu** [comment]
I found the following issue at the SAI repository: https://github.com/Aefyr/SAI/issues/91

Possibly it might help in diagnosing the FireHD 8 2018 issue.

---

Regarding arrow keys and the keyboard, possibly related: https://github.com/nesbox/TIC-80/issues/2292

**Reply by kuromame96** [reply]
I'll check the repository you suggested.
Thank you very much!

**Comment by nesbox** [comment]
Also, you can try to install fdroid version https://f-droid.org/ru/packages/com.nesbox.tic/

**Reply by kuromame96** [reply]
It seems that TIC80 version 1.1 cannot be installed on fdroid.
When I try to install TIC80 with fdroid, it installs version 1.0.

I tried downloading the apk file directly from the URL you posted and installed it, but the installation failed.
Looking at the fdroid TIC80 page you posted, it seems that TIC80 version 1.1 is not compatible with armeabi-v7a.

Apparently this is the cause.

My Fire HD 8 seems to be equipped with a 32-bit CPU that does not support arm64-v8a.
Is it your policy that TIC80 will not support 32-bit CPUs in the future?

**Reply by sogaiu** [reply]
I don't have any answers for you, but looking through [one of the Android-related build files](https://github.com/nesbox/TIC-80/blob/8c5f8b75e290378a6110f54a0d8d8471942b5a1b/build/android/app/build.gradle#L20-L22), I see:

```
        ndk {
            abiFilters 'arm64-v8a'
        }
```

Looking at [this](https://developer.android.com/ndk/guides/abis#gc) and [this](https://developer.android.com/ndk/guides/abis#sa), I wonder whether getting 32-bit support for the current case is a matter of tweaking the line...

**Reply by kuromame96** [reply]
It means that if you rewrite that line, you may be able to get 32bit support.
I'll try building it myself.

**Reply by sogaiu** [reply]
@kuromame96 Please let us know the results of your building attempt.

**Reply by kuromame96** [reply]
Still preparing to build.

The software required for installation is presented to me and I am currently installing it in order.

The installed software is not recognized for some reason if it is left as is, so I put it in the folder that contains Platform-Tools.

If you do this, it will somehow recognize you.

I'm using pretty aggressive methods to prepare for the build, so it's doubtful whether it will be successful or not.

**Reply by sogaiu** [reply, marked as answer]
Thanks for the status update.

I tried here [1] and looked inside the resulting `app-release.apk` file:

```
Archive:  ../app-release.apk
  inflating: META-INF/com/android/build/gradle/app-metadata.properties  
  inflating: classes.dex             
 extracting: lib/arm64-v8a/libtic80.so  
 extracting: lib/armeabi-v7a/libtic80.so  
  inflating: AndroidManifest.xml     
 extracting: res/CG.png              
 extracting: res/D2.png              
 extracting: res/SD.png              
 extracting: res/jy.png              
 extracting: res/u3.png              
 extracting: resources.arsc          
```

I see a file path named `lib/armeabi-v7a/libtic80.so` inside.

I don't have a device to test with though (^^;

---

[1] I followed something close to the following steps on an Ubuntu machine:

```
## basic tic-80 dependencies
sudo apt-get install git cmake ruby-full libgles1-mesa-dev libglu-dev -y

## for android
sudo apt-get install sdkmanager

## usual TIC-80 steps
git clone https://github.com/nesbox/TIC-80
cd TIC-80
git submodule init
git submodule update

## android-specific steps
cd build/android

### user-oriented android-sdk installation setup
mkdir $HOME/android-sdk
export ANDROID_SDK_ROOT=$HOME/android-sdk
yes | sdkmanager --licenses

### edit app/build.gradle so that there is a line with: abiFilters 'arm64-v8a', 'armeabi-v7a'

### build -- this step seems to automatically install necessary pieces (e.g. ndk)
./gradle assembleRelease

### copy resulting .apk file to /tmp
cp app/build/outputs/apk/release/app-release.apk /tmp
```

**Reply by kuromame96** [reply]
The steps to build using Docker in the TIC80 readme were difficult for me.
Thank you for writing the build procedure on Ubuntu.
I'm currently using Windows, so I'd like to install Ubuntu and try building.

**Reply by nesbox** [reply]
No need to set the build environment, GitHub Actions can build it for you https://github.com/nesbox/TIC-80/actions
Just make a fork of the repo, change the code, and push.

**Reply by kuromame96** [reply]
i will try!
thank you!

**Reply by kuromame96** [reply]
Before using GitHub Actions, I first created an Ubuntu environment.

As a result, the build was successful!

I installed the latest version of Ubuntu, so the software I needed was from Ubuntu 18.04.
Other than that, I followed the steps taught by @sogaiu and succeeded!
Thanks to @sogaiu for the steps to build Android on Ubuntu!

However, there was no need to rewrite app/build.gradle.
'armeabi-v7a' was already added to the abiFilters item.
Perhaps the developer saw my discussion and fixed it.
However, I wanted the Pro version quickly, so building it myself was worth it.

Thank you for taking my opinion into account!
